File: ScaleformUI_Lua/src/base/BreadcrumbsHandler.lua
BreadcrumbsHandler = setmetatable({
    breadcrumbs = {}
}, BreadcrumbsHandler)

function BreadcrumbsHandler:Count()
    return #self.breadcrumbs
end

function BreadcrumbsHandler:CurrentDepth()
    return #self.breadcrumbs -- needed? lua handles arrays from 1 not 0.. so..who knows..
end

function BreadcrumbsHandler:PreviousMenu()
    return self.breadcrumbs[#self.breadcrumbs]
end

function BreadcrumbsHandler:Forward(menu)
    table.insert(self.breadcrumbs, menu)
end

function BreadcrumbsHandler:Clear()
    local count = #self.breadcrumbs
    for i=0, count do self.breadcrumbs[i]=nil end
    self.breadcrumbs = {}
end

function BreadcrumbsHandler:Backwards()
    table.remove(self.breadcrumbs, #self.breadcrumbs)
end

File: ScaleformUI_Csharp/Menus/BreadcrumbsHandler.cs
namespace ScaleformUI.Menus
{
    public static class BreadcrumbsHandler
    {
        private static readonly List<Tuple<MenuBase, dynamic>> breadcrumbs = new List<Tuple<MenuBase, dynamic>>();
        internal static int Count => breadcrumbs.Count;
        internal static int CurrentDepth => breadcrumbs.Count == 0 ? 0 : breadcrumbs.Count - 1;
        public static MenuBase PreviousMenu => breadcrumbs[CurrentDepth - 1].Item1;
        public static bool SwitchInProgress = false;

        internal static void Forward(MenuBase menu, dynamic data)
        {
            breadcrumbs.Add(new Tuple<MenuBase, dynamic>(menu, data));
        }

        internal static void Clear()
        {
            breadcrumbs.Clear();
        }

        internal static void Backwards()
        {
            breadcrumbs.RemoveAt(CurrentDepth);
        }
    }
}
